ludwig/data/negative_sampling.py

```diff
@@ -0,0 +1,106 @@
+import logging
+import time
+from typing import Any, List, Tuple
+
+import numpy as np
+import pandas as pd
+import scipy
+
+from ludwig.utils.types import DataFrame
+
+
+def _negative_sample_user(interaction_row: np.array, neg_pos_ratio: int, extra_samples: int) -> Tuple[List[int], int]:
+    """Returns a list of negative item indices for given user-item interactions.
+
+    If there are not enough negative items, takes all of them and adds the difference to the extra_samples
+    otherwise, samples with replacement.
+
+    Params:
+        interaction_row: user-item interaction row
+        neg_pos_ratio: number of negative samples per positive sample
+        extra_samples: number of additional samples to add to the negative sample list
+    Returns:
+        Tuple of list of negative item indices and number of extra samples
+    """
+    # Find all items that are not interacted with by the user
+    neg_items = np.where(interaction_row == 0)[1]
+    available_samples = len(neg_items)
+
+    # Randomly sample negative items
+    npos = interaction_row.shape[1] - len(neg_items)
+    samples_required = npos * neg_pos_ratio + extra_samples
+    should_sample = samples_required <= available_samples
+
+    neg_items = np.random.choice(neg_items, samples_required, replace=False) if should_sample else neg_items
+
+    return neg_items.tolist(), max(0, samples_required - available_samples)
```

Comment:
Wondering if we can be more opinionated here and actually make our negative samples even more specific. Instead of sampling from the entire world of negative labels, what if we only create negative labels from other positive labels? 

For e.g.,

user A: [1, 2, 3]
user B: [1, 2, 6]
user C: [1, 2, 5, 8]
All items: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In this case, as I understand it, for user A, neg_items will be [4, 5, 6, 7, 8, 9, 10] and we will randomly sample from this list. However, if items 7, 9, and 10 have never seen any positive interaction, I probably wouldn't even consider them as negative candidates for any user. I'd instead, look at all of the items that were positive across all users (in this case, [1, 2, 3, 5, 6, 8] and then sample from the diff between what user A interacted with and everything that all users interacted with = [5, 6, 8].

The advantage of this is that it really helps when the item space is large but actual positive interactions happen with just a smaller subset of the item space. The disadvantage is that doesn't really support discovery outside of what's already been interacted with by at least 1 user. 

I've seen this opinionated approach work pretty well on average, but might be consider when to use what strategy for negative sampling. 

What do you think?

Reply:
I guess the case I'm making is that maybe in the future, we can create different negative sampling strategies because I assume that some will be more performant / better for certain recommendation-system use cases.

Reply:
I really like this idea. It could be worth adding an abstract `NegativeSamplingStrategy` class. Implementations can then differ in how they select negative samples. E.g. `FromAll(NegativeSamplingStrategy)` or `FromAllPositive(NegativeSamplingStrategy)`

Reply:
@jppgks I would definitely be in support of a NegativeSamplingStrategy abstract class implementation and abstraction for strategies :) I have a sense that we may discover even more ideas on the way for negative sampling strategies. 

Could we do your proposed implementation as a part of this PR, or would that be too much additional work?

Reply:
@jppgks I can already think of a third strategy that combines the two strategies with some sort of balance parameter 'ratio', where for e.g., 80% of the negative samples come from FromAllPositive and 20% comes from FromAll (and we ensure no overlap between the two sets) to ensure a way to balance what's been interacted with some sort of discovery element.

Reply:
In the future, we can come up with a heuristic to pick these strategies based on the dataset profile.